Encode credentials into qr codes with cborld

kezike opened the pull request:
Encoding credentials into QR Codes with `CBOR-LD` via Digital Bazaar's `vpqr` library

Here is a sample credential:
```
{
  "@context": [
    "https://www.w3.org/2018/credentials/v1",
    "https://w3id.org/security/suites/ed25519-2020/v1",
    "https://w3id.org/dcc/v1"
  ],
  "id": "https://digitalcredentials.github.io/samples/certificate/1fe91f0f-4c64-48c8-bfc8-7132f75776fe/",
  "type": [
    "VerifiableCredential",
    "LearningCredential"
  ],
  "issuer": {
    "type": "Issuer",
    "id": "did:web:digitalcredentials.github.io",
    "name": "Sample Issuer",
    "url": "https://digitalcredentials.github.io/samples/"
  },
  "issuanceDate": "2021-01-19T18:22:34.772810+00:00",
  "credentialSubject": {
    "type": "Person",
    "id": "did:example:456",
    "name": "Percy",
    "hasCredential": {
      "type": [
        "EducationalOccupationalCredential",
        "ProgramCompletionCredential"
      ],
      "name": "DCC Program Completion Credential",
      "description": "Awarded on completion of the digital credential program",
      "awardedOnCompletionOf": {
        "type": "EducationalOccupationalProgram",
        "identifier": "program-v1:Sample",
        "name": "Digital Credential Program",
        "description": "Educational program teaching how work with digital credentials",
        "numberOfCredits": {
          "value": "1"
        },
        "startDate": "",
        "endDate": ""
      }
    }
  }
}
```

and its associated QR Code
![qr-code](https://user-images.githubusercontent.com/8096163/139557917-34e83877-89d5-4fc9-8b62-06097eab7629.gif)

dmitrizagidulin commented on src/utils/codecs.ts:
```diff
@@ -0,0 +1,33 @@
+const didContext = require('@digitalcredentials/did-context');
+const ed25519 = require('ed25519-signature-2020-context');
+const DccContextV1Url = "https://w3id.org/dcc/v1";
+import { DccContextV1 } from "./contexts";
```
Any reason to import from a local copy, instead of from the NPM module? (see https://github.com/digitalcredentials/dcc#usage for example).

kezike replied:
Interesting - I hadn't seen this usage of `dcc-context`. I was using [this sample code](https://github.com/digitalcredentials/sign-and-verify-core/blob/master/src/common.ts#L3) from `sign-and-verify-core` as a reference point. However, when I try to install `dcc-context` in the manner suggested here, I came across this error despite having a later version of Node.js than the error suggests:

![Screen Shot 2021-10-31 at 7 53 34 PM](https://user-images.githubusercontent.com/8096163/139605663-adb298d1-6835-48fe-8ddd-ad2ba1c069cf.png)

dmitrizagidulin replied:
@kezike oh weird.. 
`dcc-context` module does not have an "engine" directive in its package.json; I wonder where that error is coming from!

Let's debug -- what npm and Node version are you using? Do you have any further details on that error (like in the npm error log)?

dmitrizagidulin replied:
The wallet uses the dcc-context package -- you can see the usage here: https://github.com/digitalcredentials/learner-credential-wallet/blob/main/app/lib/documentLoader.ts#L3
(And it does not result in any npm version error).

kezike replied:
Never mind - I was able to resolve this issue after a system restart (computer's been a bit wonky lately)!

dmitrizagidulin commented on src/modules.d.ts:
```diff
@@ -0,0 +1,3 @@
+// TODO: The DCC fork of vpq is not working at the moment
+// declare module '@digitalcredentials/vpqr';
```
I enabled Issues on the DCC fork -- do you mind opening an issue with more details?

kezike replied:
Great, thanks! Here is the issue: https://github.com/digitalcredentials/vpqr/issues/2

dmitrizagidulin reviewed (approved):
:+1: 
Thanks @kezike!

kezike commented:
@dmitrizagidulin In this PR, I am using `vpqr` to encode the VC directly. Should I be placing the VC into a VP envelope before using this library or is it able to encode VCs as expected. I assume the library accepts and treats both formats the same way, but I just want to be sure.

dmitrizagidulin commented:
> Should I be placing the VC into a VP envelope before using this library or is it able to encode VCs as expected. I assume the library accepts and treats both formats the same way, but I just want to be sure.

You definitely want to place the VC into a VP envelope. In general, un-wrapped VCs should only exist inside the wallet / VC storage; they shouldn't really be encountered in the wild.

kezike commented:
> > Should I be placing the VC into a VP envelope before using this library or is it able to encode VCs as expected. I assume the library accepts and treats both formats the same way, but I just want to be sure.
> 
> You definitely want to place the VC into a VP envelope. In general, un-wrapped VCs should only exist inside the wallet / VC storage; they shouldn't really be encountered in the wild.

Makes sense! I have updated the PR to package a VC into an unsigned VP before encoding it into a QR Code URL. I have also implemented the `ProvePresentationRequest` method to interface with the `/prove/presentations` endpoint of the `sign-and-verify` API, but this will only become relevant after we host a `sign-and-verify` instance.
